### PR TITLE
Localize jsonreader infrastructure.

### PR DESCRIPTION
**Pull request by BillyONeal**
* Change type_name and friends to return LocalizedString instead of StringView as this now requires formatting.
* StringDeserializer and ArrayDeserializer are "nerfed" in that they can't provide type_name; this now requires separate derivation because figuring out what type_name is must be deferred until after the current locale is loaded.

### CODE REVIEW

File: include/vcpkg/base/jsonreader.h [hunk 1]
```diff
@@ -242,145 +247,115 @@ namespace vcpkg::Json
...
-
-        static BooleanDeserializer instance;
+        virtual LocalizedString type_name() const override;
+        virtual Optional<bool> visit_boolean(Reader&, bool b) const override;
+        static const BooleanDeserializer instance;
     };
 
     template<class Underlying>
-    struct ArrayDeserializer final : IDeserializer<std::vector<typename Underlying::type>>
+    struct ArrayDeserializer : IDeserializer<std::vector<typename Underlying::type>>
```
Comment (soroshsabz) [inline include/vcpkg/base/jsonreader.h:319]: Why remove `final` from this struct?

File: include/vcpkg/base/jsonreader.h [hunk 2]
```diff
@@ -17,21 +17,21 @@ namespace vcpkg::Json
     struct IDeserializer
     {
         using type = Type;
-        virtual StringView type_name() const = 0;
+        virtual LocalizedString type_name() const = 0;
```
Comment (soroshsabz) [inline include/vcpkg/base/jsonreader.h:20]: Are you sure to change StringView to LocalizedString? I think change StringView to LocalizedStringView

File: include/vcpkg/base/jsonreader.h (hunk 1 above)
Comment (BillyONeal) [reply include/vcpkg/base/jsonreader.h:319]: It is now abstract.

**Review by JavierMatosD — APPROVED** [review]

**Comment by BillyONeal** [comment]
@JavierMatosD Thanks!

**Pull request merged**